git-hub

```diff
@@ -816,61 +813,113 @@ class CloneCmd (object):
+            git_remotes = git('remote', 'show', '-n').split('\n')
+            remote = args.upstreamremote
+            if remote not in git_remotes:
+                die("No REPO specified, nor does `{}` remote exist", remote)
+            git_url = git('remote', 'get-url', '--', remote)
+            (urltype, proj) = cls.parse_repo(git_url)
+        (repo, upstream, forked) = cls.maybe_fork(proj)
+        cls.setup_dest(parser, args, urltype, proj, repo, upstream, forked, None)
+
+    @classmethod
+    def setup_dest(cls, parser, args, urltype, proj, repo, upstream, forked, dest):
+        personal = False
+        if upstream is None:
+            if args.upstreamremote != args.forkremote:
+                die('You are setting up with a personal repository as upstream, '
+                        '`--upstreamremote` and `--forkremote` must match')
+            upstream = proj
+            personal = True
         triangular = cls.check_triangular(args.triangular
                 if args.triangular is not None else config.triangular)
-        if triangular and not upstream:
-            parser.error("Can't use triangular workflow without "
-                    "an upstream repo")
-        url = repo['parent'][urltype] if triangular else repo[urltype]
+        if personal and triangular:
+            warnf('You are setting up with a personal repository as upstream, '
+                        'forcing `--no-triangular` mode')
+            triangular = False
+        if triangular:
+            url = repo['parent'][urltype]
+            remote = args.upstreamremote
+        else:
+            url = repo[urltype]
+            remote = args.forkremote
         validate_url(url, urltype)
-        remote = args.upstreamremote if triangular else args.forkremote
-        # It's complicated to allow the user to use the --origin option, there
-        # is enough complexity with --upstreamremote and --forkremote, so ask
-        # the user to use those instead
-        for a in args.unknown_args:
-            if a in ('-o', '--origin') or a.startswith('-o', '--origin='):
-                die("Please use --forkremote or --upstreamremote to name your "
-                    "remotes instead of using Git's `{}` option!", a)
-        # If we just forked the repo, GitHub might still be doing the actual
-        # fork, so cloning could fail temporarily. See
-        # https://github.com/sociomantic-tsunami/git-hub/issues/214
-        cls.git_retry_if(not args.triangular and forked,
-                'clone',
-                args.unknown_args + ['--origin', remote, '--', url, dest],
-                'Cloning {} to {}'.format(url, dest))
-        if not upstream:
-            # Not a forked repository, nothing else to do
-            return
-        # Complete the repository setup
-        os.chdir(dest)
-        fetchremote = args.forkremote if triangular else args.upstreamremote
-        remote_url = repo['parent'][urltype]
+        if dest is not None:
```

Comment:
This code will only ever used when invoking this function from the `clone` command, right? If this is correct, wouldn't it make more sense to just move this code to the `clone` command?

If not, wouldn't it make sense to rename `dest` to `clone_to` or something like that to make it more explicit that a `clone` can happen here?

In general I'm not a fan of such a large method, so If you find any way to factor out some parts, it would be greatly appreciated. Maybe just remove some small parts as independent methods and then write slightly longer `run` methods for `setup-fork` and `clone` would make more sense to me.